### PR TITLE
Drop readable unique pending

### PR DESCRIPTION
**Pull request by ScottDugas**

### CODE REVIEW

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java
Comment (jjezra) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java:596]:
```suggestion
      if (state.index.isUnique()) {
            throw new RecordCoreException("Lucene index does not support uniqueness constraint");
      }
```

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMergeTest.java
Comment (jjezra) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMergeTest.java:257]: Would this index ever become readable?

**Review by jjezra — COMMENTED** [review]
I am reviewing this PR while in Draft mode (vacation it is), my apologizes for any premature nit-picking.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreUniqueIndexTest.java
Comment (alecgrieser) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreUniqueIndexTest.java:721]: It might be worth highlighting that this method is the only one that _doesn't_ delegate to the value index maintainer, resulting in `clearUniquenessViolations` being a no-op. I think it's pretty clear form the index name that that's the intention, but highlighting that (somewhat subtle) difference is probably good for future maintainers trying to understand how it achieves that goal.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreUniqueIndexTest.java
Comment (alecgrieser) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreUniqueIndexTest.java:414]: Is it worth adding a test of what happens with the `NO_UNIQUE_CLEAR_INDEX_TYPE` type and we don't bump the version? I think I'd expect it to act like the index doesn't have any uniqueness constraints, but we'd still be able to find some by scanning the index violations subspace

**Review by alecgrieser — CHANGES_REQUESTED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java
Comment (ScottDugas) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java:1523]: Yeah, I filed another issue for that: #2919 and will fix separately.

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreUniqueIndexTest.java
Comment (ScottDugas) [reply fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreUniqueIndexTest.java:721]: Good call. I added a javadoc. And since this is now using the default implementation, instead of defining its own method to throw, I renamed it.

**Review by ScottDugas — COMMENTED** [review]
I addressed all the comment except the one that I put in the separate PR: [2920: Fixes #2919: scope the IndexUniquenessCommitCheck by subpsace](https://github.com/FoundationDB/fdb-record-layer/pull/2920)

Once #2920 is merged, I will merge it into this, and adjust as appropriate.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContext.java
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContext.java:803]: I'd be a little concerned about a `ConcurrentModificationException`s here, if `removeCommitCheck` attempts to remove an element while the `stream` is still going over it. The alternative is probably to do something like (1) collect the filtered commit checks into a list and then (2) remove them only once the iteration is done.

**Review by alecgrieser — DISMISSED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContext.java
Comment (ScottDugas) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContext.java:803]: `removeCommitCheck` and `removeCommitChecks` are both `synchronized`, so it won't be able to remove the check while it is being iterated over, unless I'm missing something

**Review by alecgrieser — APPROVED** [review]

**Pull request merged**
